docs/syntax-jsx-support.md

```diff
@@ -0,0 +1,338 @@
+- Avoid hard-coding any specific UI library into the language by desugaring to a **user-provided factory function**.
+
+### Why JSX belongs at the language level (and is not “React syntax”)
+
+A common concern is that JSX is “too specific” to React or UI frameworks to warrant native syntax. This RFC argues the opposite: **JSX is a general-purpose notation for declarative, structured function calls**, and the runtime meaning is entirely controlled by the `jsx` factory function chosen by the host/project.
+
+In other words, JSX is not “React in the language”; it is:
+
+- A compact syntax for constructing nested call trees (`jsx(tag, props, ...children)`).
+- A convenient way to bind named parameters via record-like props (`{ key = value }`).
+- A structured form that is easier for humans and tools to read than deeply nested function calls.
+
+This makes JSX useful beyond UI, including data-driven scene graphs, entity graphs, or declarative configuration trees.
+
+## Design
+
+### Goals
+
+- **No transpilation step**: JSX is parsed directly by Luau’s parser when enabled.
+- **Opt-in**: no behavior change for `.lua` / `.luau` unless explicitly enabled.
+- **Library-agnostic**: JSX lowers to an ordinary Luau call (`jsx(...)`) so projects can bind it to Roact/React/etc.
+- **Minimize grammar ambiguity** with existing and future Luau syntax.
+
+### Non-goals (initially)
+
+These are intentionally excluded from the initial proposal to reduce ambiguity and implementation risk:
+
+- Raw text children (e.g. `<Foo>hello</Foo>`)
+- Fragments (e.g. `<>...</>`)
+- Spread attributes (e.g. `<Foo {...props} />`)
+- Namespaced attributes (e.g. `on:click=...`)
+- JSX comments syntax
+
+These can be added via follow-on RFCs once the core parsing approach proves robust.
+
+### Opt-in surface area
+
+**File extension**: `.luax` indicates JSX may appear in the file.
+
+**Feature flag**: The syntax must be gated by a flag (e.g. `FFlag::LuauJsx`) so it can ship experimentally and be rolled out carefully.
+
+### Syntax
+
+When enabled, an additional `simpleexp` form is introduced:
+
+- `jsxElement ::= '<' jsxTagName jsxAttributes ( '/>' | '>' jsxChildren '</' jsxTagName '>' )`
+- `jsxTagName ::= Name { '.' Name }`
+- `jsxAttributes ::= { jsxAttribute }`
+- `jsxAttribute ::= Name [ '=' ( String | '{' expr '}' ) ]`
+- `jsxChildren ::= { jsxChild }`
+- `jsxChild ::= jsxElement | '{' expr '}'`
+
+Notes:
+
+- `{ expr }` is used for embedding arbitrary Luau expressions in children and attribute values.
+- Attribute shorthand (`<Foo disabled />`) is allowed and is equivalent to `disabled={true}`.
+
+### Desugaring
+
+JSX constructs are desugared in the parser into an ordinary Luau function call:
+
+#### Tag lowering
+
+- If the tag begins with a lowercase identifier and has no dots (e.g. `<frame />`), it lowers to a string tag: `"frame"`.
+- Otherwise it lowers to an identifier/dotted expression (e.g. `<Foo.Bar />` lowers to `Foo.Bar`).
```

Comment:
This feels very non-obvious, and like it would be a constant footgun for using this feature. Why is this the semantics you want to propose? Why are we not proposing a single, consistent semantics for tags?

Reply:
This is the convention of JSX, which alone is not justification for how we should do it, but does give the benefit of skill transferability for developers already familiar with it.

One counter argument against the convention is it makes mapping strings to intrinsic elements harder to manage as theres no guarantee your instances (like Roblox's instances which typically use pascal case) which will usually require an implementer to provide a mapping from strings to intrinsic element names.

Again the largest argument against this is to leverage familiar widely adopted conventions.